crates/burn-jit/src/fusion/on_write/builder.rs

```diff
@@ -16,10 +16,40 @@ use cubecl::ir::Elem;
 
 /// Fused element wise operations that are normally memory bound.
 pub(crate) struct FuseOnWriteBuilder {
-    builder: FuseOnWriteTraceBuilder,
+    builder: TryFuseBuilder,
     current_output_shape: Vec<usize>,
     status: OptimizationStatus,
-    num_added: usize,
+    num_ops: usize,
+    num_tensors: usize,
+    max_bindings: u32,
+}
+
+struct TryFuseBuilder {
+    builder: FuseOnWriteTraceBuilder,
+    max_bindings: u32,
+}
+
+impl TryFuseBuilder {
+    fn new(max_bindings: u32) -> Self {
+        Self {
+            builder: FuseOnWriteTraceBuilder::new(),
+            max_bindings,
+        }
+    }
+
+    fn register(&mut self, build: impl FnOnce(&mut FuseOnWriteTraceBuilder)) -> bool {
+        let mut cloned = self.builder.clone();
+        build(&mut cloned);
+        if cloned.estimate_bindings() > self.max_bindings {
+            return false;
+        }
+        self.builder = cloned;
+        true
```

Comment:
cloning every time is a bit wasteful. I would only do it if `builder_current_state.estimate_bindings() + new_op.nodes() > self.max_bindings`

Reply:
That doesn't really work - we don't know how many nodes someone is adding in the build function. And we can't easily keep track of it as adding them mutates the builder.

estimate_bindings() also seems more expensive than a clone in the first place?

I've at least added a fast path for the first operation which should be common. If that's not enough might need to rethink how these operations are added so we can count them more easily.